### DISCUSSION TITLE
React Native managing multiple screens with same store

### DISCUSSION THREAD

**Discussion by murat-mehmet** [mobxjs/mobx#3054] [Q&A] [answered]
In React Native we are using a mobX store to keep product detail data. However the product detail page can be navigated multiple times by clicking another product under Similar Products tab. So in general I need to open the same 'product detail' screen multiple times with different data. I looked very hardly to find an appropriate way of doing this with mobX. There are many nested components on the page so I don't want to pass the navigated product data to every component, there will be no point in using mobx.

One idea is to use a nested Provider in this page and create and pass a new store to this provider, so every product detail page will share same logic and will be provided separate stores. When testing this seems to work great but while developing on this page, hot-reloading keeps complaining about 'The set of provided stores has changed'.

So am I on right track? Is there any other way of doing this more clear? Is it ok to override Provider store with same key in sub-screens like this?

**Comment by mweststrate** [comment]
Sounds fine to mee. Note that you don't need to use mobx's Provider anymore
as it is deprecated, since React's createContext provides largely the same
functionality.

**Comment by murat-mehmet** [comment, marked as answer]
Ok thanks for the response. This is the decorator I prepared for such screens.
```
const overrideStore = (...storeKeys) => (WrappedComponent) => {
  return function (props) {
    let overrides = {};
    const stores = React.useContext(MobXProviderContext)
    storeKeys.forEach(storeKey => {
      if (!stores[storeKey])
        throw new Error(`'${storeKey}' store does not exist`)
      const StoreType = stores[storeKey].constructor;
      overrides[storeKey] = new StoreType(stores);
    })
    return (
        <Provider {...overrides}>
          <WrappedComponent {...props} />
        </Provider>
    )
  };
}
```
usage:
```
@overrideStore('product')
@inject('product')
@observer
class ProductDetail extends Component {
...
```
